lib/index.js

```diff
@@ -0,0 +1,233 @@
+  // *always* be set)
+
+  // FIXME: should the default `MAX_ELECTOR_COUNT` be removed given the
+  // requirements around recovery mode?  Specifically, *all* recovery mode mode
+  // electors (whatever number) *must* be electors as well.
+  // if so, the ledgerConfiguration validator must require `electorCount`
+  const electorCount = ledgerConfiguration.electorCount || MAX_ELECTOR_COUNT;
+
+  // TODO: could optimize by only sorting tied electors if helpful
+  /*
+  // fill positions
+  let idx = -1;
+  for(let i = 0; i < electorCount; ++i) {
+    if(electors[i].weight > electors[i + 1].weight) {
+      idx = i;
+    }
+  }
+  // fill positions with non-tied electors
+  const positions = electors.slice(0, idx + 1);
+  if(positions.length < electorCount) {
+    // get tied electors
+    const tied = electors.filter(
+      e => e.weight === electors[idx + 1].weight);
+    // TODO: sort tied electors
+  }
+  }*/
+
+  const {blockHash} = latestBlockSummary.eventBlock.meta;
+
+  // break ties via sorting
+  electors.sort((a, b) => {
+    // 1. sort descending by weight
+    if(a.weight !== b.weight) {
+      // FIXME: with current weights, this prevents elector cycling
+      //   if commented out, will force elector cycling, needs adjustment
+      return b.weight - a.weight;
+    }
+
+    // FIXME: when mixing in data, why not `xor` instead of sha-256?
+
+    // generate and cache hashes
+    // the hash of the previous block is combined with the elector id to
+    // prevent any elector from *always* being sorted to the top
+    a.hash = a.hash || _sha256(blockHash + _sha256(a.id));
+    b.hash = b.hash || _sha256(blockHash + _sha256(b.id));
+
+    // 2. sort by hash
+    return a.hash.localeCompare(b.hash);
+  });
+
+  // select first `electorCount` electors
+  electors = electors.slice(0, electorCount);
+
+  // TODO: if there were no electors chosen or insufficient electors,
+  // add electors from config
+
+  // only include `id`
+  electors = electors.map(e => ({id: e.id}));
+
+  // reduce electors to highest multiple of `3f + 1`, i.e.
+  // `electors.length % 3 === 1` or electors < 4 ... electors MUST be a
+  // multiple of `3f + 1` for BFT or 1 for trivial dictator case
+  while(electors.length > 1 && (electors.length % 3 !== 1)) {
+    electors.pop();
+  }
+
+  return electors;
+};
+
+// it is useful to override this function in tests
+api._computeElectorsForRecoveryMode = (
+  {blockHeight, electors, recoveryElectors}) => {
+  // FIXME: is this what we're doing here?
+  return recoveryElectors;
+
+  // FIXME: remove
+  // for this implementation, simply pick the first recovery elector
+  // recoveryElectors.sort((a, b) => a.id.localeCompare(b.id));
+  // return [recoveryElectors[0]];
+};
+
+// it is useful to override this function in tests
+api._computeRecoveryElectors = (
+  {blockHeight, electors, electorPoolElectors, f}) => {
+  // FIXME: is `f` a factor here?
+  // FIXME: if elector.type is a thing, filter by type?
```

Comment:
Yes, we should have some kind of type that is in the elector pool document that says an elector is to be used for recovery mode ... and we'll also want a `guarantor` or ("always an elector") type that gives preference to certain electors such that they are always chosen (in regular/non-recovery mode).

Reply:
recovery mode electors must always be electors right?  So that covers the `guarantor` type.  So then there is evidently a use case for specifying electors that are *not* recovery mode electors, what should that type be?

Reply:
> recovery mode electors must always be electors right? So that covers the guarantor type. 

This is not the same thing. I'm saying that there are a set of electors that are *always* chosen for a particular block. These are "guarantors". The set of recovery electors is likely a subset of those, but that's a separate issue.

> So then there is evidently a use case for specifying electors that are not recovery mode electors, what should that type be?

Electors that are recovery mode electors should have a recovery type. This is different from the guarantor type. Some may have both.

Does that make sense?